### PR TITLE
Add UnitTest for DeepES

### PR DESCRIPTION
**Pull request by rical730**

### CODE REVIEW

File: deepes/test/src/test_utils.cc [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+#include "gtest/gtest.h"
+#include <vector>
+#include "utils.h"
+
+namespace DeepES {
+
+// The fixture for testing class or file Utils.
+class UtilsTest : public ::testing::Test {
+ protected:
+  // You can remove any or all of the following functions if their bodies would
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/src/test_utils.cc]: Please remove empty functions.

File: deepes/test/src/test_utils.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+};
+
+// Tests that the Utils::compute_centered_rank() method.
+TEST_F(UtilsTest, Method_compute_centered_ranks) {
+  float a[5] = {9.0, 8.0, 7.0, 6.0, 5.0};
+  std::vector<float> reward_vec(a, a+5);
+  EXPECT_EQ(compute_centered_ranks(reward_vec), true);
+}
+
+// // Tests that Utils another method Xyz.
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/src/test_utils.cc]: remove the comment here.

File: deepes/test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+endif()
+
+# if (WITH_TORCH)
+#     list(APPEND CMAKE_PREFIX_PATH "../libtorch")
+#     find_package(Torch REQUIRED ON)
+#     include_directories("../include/torch")
+#     file(GLOB lib_src "../src/torch/*.cc")
+# endif()
+
+# file(GLOB demo_src "../demo/*.cc" "../demo/*.cpp")
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/CMakeLists.txt]: remove the commented line.

File: deepes/test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+file(GLOB core_src "../src/*.cc" "../src/*.cpp" "../benchmark/*.cc")
+file(GLOB test_src "../test/src/*.cc")
+
+# include_directories("../demo")
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/CMakeLists.txt]: remove the commented line.

File: deepes/test/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+include_directories("../include")
+include_directories("../benchmark")
+include_directories("../test/include")
+# link_directories("/usr/lib/x86_64-linux-gnu/")
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/CMakeLists.txt]: remove the commented line.

**Comment by TomorrowIsAnOtherDay** [comment]
@rical730  Please add the unit test for optimizers with the following test cases:
- create different optimizers through the optimizer_factory
- pass the gradient with same param_name but with different param_size
- given gradient and current parameters, check if the optimizer return the ideal result.

File: deepes/test/src/test_utils.cc (hunk 1 above)
Comment (rical730) [reply deepes/test/src/test_utils.cc]: done

File: deepes/test/src/test_utils.cc (hunk 2 above)
Comment (rical730) [reply deepes/test/src/test_utils.cc]: done

File: deepes/test/CMakeLists.txt (hunk 3 above)
Comment (rical730) [reply deepes/test/CMakeLists.txt]: done

File: deepes/test/CMakeLists.txt (hunk 4 above)
Comment (rical730) [reply deepes/test/CMakeLists.txt]: done

File: deepes/test/CMakeLists.txt (hunk 5 above)
Comment (rical730) [reply deepes/test/CMakeLists.txt]: done

**Comment by rical730** [comment]
> @rical730 Please add the unit test for optimizers with the following test cases:
> 
> * create different optimizers through the optimizer_factory
> * pass the gradient with same param_name but with different param_size
> * given gradient and current parameters, check if the optimizer return the ideal result.

will be done in next PR

File: deepes/test/run_test.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
+
+#----------------protobuf-------------#
+rm deepes.proto
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/run_test.sh]: remove this line, as the following line will copy a new deepes.proto to the current directory.

Comment (rical730) [reply deepes/test/run_test.sh]: done

File: deepes/test/src/optimizers_test.cc [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	float sgd_new[10]  = { 0.01199242, 0.0       , 0.0344831 , 0.05776198, 0.04206595, 0.00973154, 0.09637211,-0.03477474, 0.014892306, 0.03129495};
+
+	EXPECT_TRUE(optimizer->update(sgd_wei, sgd_grad, 10, "test"));
+	for (int i = 0; i < 10; ++i) {
+		EXPECT_FLOAT_EQ(sgd_new[i], sgd_wei[i]) << " i: " << i ;
+	}
+	EXPECT_FALSE(optimizer->update(sgd_wei, sgd_grad, 9, "test"));
+}
+
+// Tests that the Optimizers::compute_centered_rank() method.
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/src/optimizers_test.cc]: remove the unrelated comment.

File: deepes/test/src/optimizers_test.cc [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+#include "gtest/gtest.h"
+#include <vector>
+#include "optimizer.h"
+#include "sgd_optimizer.h"
+#include "adam_optimizer.h"
+#include <memory>
+
+namespace DeepES {
+
+// Tests that the Optimizers::compute_centered_rank() method.
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/src/optimizers_test.cc]: remove the unrelated comment.

File: deepes/test/src/optimizers_test.cc (hunk 7 above)
Comment (rical730) [reply deepes/test/src/optimizers_test.cc]: done

File: deepes/test/src/optimizers_test.cc (hunk 8 above)
Comment (rical730) [reply deepes/test/src/optimizers_test.cc]: done

**Comment by rical730** [comment]
> @rical730 Please add the unit test for optimizers with the following test cases:
> 
> * create different optimizers through the optimizer_factory
> * pass the gradient with same param_name but with different param_size
> * given gradient and current parameters, check if the optimizer return the ideal result.

done

File: deepes/test/src/optimizers_test.cc [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace DeepES {
+
+TEST(SGDOptimizersTest, Method_update) {
+    std::shared_ptr<DeepESConfig> config = std::make_shared<DeepESConfig>();
+	auto optimizer_config = config->mutable_optimizer();
+	optimizer_config->set_base_lr(1.0);
+	optimizer_config->set_type("sgd");
+	std::shared_ptr<Optimizer> optimizer = create_optimizer(config->optimizer());
+	// std::shared_ptr<Optimizer> optimizer = std::make_shared<SGDOptimizer>(1.0);
```
Comment (TomorrowIsAnOtherDay) [inline deepes/test/src/optimizers_test.cc]: remove this line.

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
Looks good to me.

**Pull request merged**